app/src/main/java/com/seugi/util/FirebaseMessageService.kt

```diff
@@ -0,0 +1,36 @@
+package com.seugi.util
+
+import android.util.Log
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+import com.seugi.data.firebasetoken.FirebaseTokenRepository
+import dagger.hilt.android.AndroidEntryPoint
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+@AndroidEntryPoint
+class FirebaseMessageService : FirebaseMessagingService() {
+
+    @Inject
+    lateinit var firebaseRepository: FirebaseTokenRepository
+
+    override fun onNewToken(token: String) {
+        super.onNewToken(token)
+
+        CoroutineScope(Dispatchers.IO).launch {
+            Log.d("TAG", "firebase Token: $token")
+            firebaseRepository.insertToken(
+                firebaseToken = token,
+            )
+        }
+    }
+
+    override fun onMessageReceived(message: RemoteMessage) {
+        super.onMessageReceived(message)
+        Log.d("TAG", "Notification Title :: ${message.notification?.title}")
+        Log.d("TAG", "Notification Body :: ${message.notification?.body}")
+        Log.d("TAG", "Notification Data :: ${message.data}")
+    }
+}
```

Comment:
해당 코드에서 CoroutineScope를 만들고 따로 관리하지 않는 것으로 보이는데, 아래와 같이 scope를 제어하는게 어떨까요?


```suggestion
@AndroidEntryPoint
class FirebaseMessageService : FirebaseMessagingService() {

    private val serviceScope = CoroutineScope(EmptyCoroutineContext)

    @Inject
    lateinit var firebaseRepository: FirebaseTokenRepository

    override fun onNewToken(token: String) {
        super.onNewToken(token)

        serviceScope.launch(Dispatchers.IO) {
            Log.d("TAG", "firebase Token: $token")
            firebaseRepository.insertToken(
                firebaseToken = token,
            )
        }
    }

    override fun onMessageReceived(message: RemoteMessage) {
        super.onMessageReceived(message)
        Log.d("TAG", "Notification Title :: ${message.notification?.title}")
        Log.d("TAG", "Notification Body :: ${message.notification?.body}")
        Log.d("TAG", "Notification Data :: ${message.data}")
    }

    override fun onDestroy() {
        super.onDestroy()
        serviceScope.cancel()
    }

}
```

Reply:
왜 따로 관리해야하나요

Reply:
현재 코드는 coroutineScope가 직접 생성하고 관리를 하지 않습니다. 이로 인해 생명주기에 맞지 않게 돌아가지 않습니다. 이를 해결하기 위해 Service 컴포넌트가 종료될 때 coroutineScope를 cancel하는 것으로 Service의 생명주기에 맞게 돌릴 수 있습니다.

Reply:
EmptyCoroutineContex이건 뭔가요

Reply:
비어있는 CoroutineContext를 만드는 것입니다.